### PR TITLE
Fix MacOS dynamic linking issue

### LINKED ISSUE #521: Dynamic linking issue on MacOS
Building binaries with dynamic linking (`linkstatic = False`) on MacOS causes them to fail at execution due to dynamic linker errors.

### Steps to reproduce

- Check out or cherry-pick 86872194170a835d2ff84e9800d71f1030be2d3e
    (That commit adds a regression test for this issue)
    The test case defines a Haskell binary with a Haskell library dependency, both linked dynamically.
    It then defines a `genrule` that executes the binary.
- Execute the following command on MacOS and observe the failure.
    ```
    $ bazel build //tests:run-bin-with-lib-dynamic
    ...
    dyld: Library not loaded: @rpath/libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.4.4.dylib.temp
      Referenced from: /private/var/tmp/_bazel_builder/d1983bacedc8743311e709889ea53957/sandbox/darwin-sandbox/6/execroot/io_tweag_rules_haskell/bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic
      Reason: image not found
    sh: line 2:  5720 Abort trap: 6           bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic
    Target //tests:run-bin-with-lib-dynamic failed to build
    ...
    ```
    (The error occurs in both cases, when run within a pure Nix shell or without).

### Additional information

- Note, that the executable tries to load a `.dylib.temp` file. To my understanding the `.temp` file is an [intermediate build product](https://github.com/tweag/rules_haskell/blob/8bc2b2c847c54f3d9f6bd5000f8deefa1cf4c995/haskell/private/actions/link.bzl#L362) and should not be linked against. Instead, the executable should link against the final `.dylib`.

    Adding `ls -R` to the test-case's `genrule` reveals that no `.dylib.temp` file exists in the execution environment:

    ```
    ...
    ./bazel-out/host/bin/tests:
    binary-with-lib-dynamic
    
    ./bazel-out/host/bin/tests/binary-with-lib-dynamic:
    binary-with-lib-dynamic
    binary-with-lib-dynamic.runfiles
    
    ./bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic.runfiles:
    io_tweag_rules_haskell
    
    ./bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic.runfiles/io_tweag_rules_haskell:
    tests
    
    ./bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic.runfiles/io_tweag_rules_haskell/tests:
    binary-with-lib-dynamic
    
    ./bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic.runfiles/io_tweag_rules_haskell/tests/binary-with-lib-dynamic:
    binary-with-lib-dynamic
    libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.4.4.dylib
    ...
    ```

- Running `otool -l` on the binary shows
    ```
    Load command 13
              cmd LC_LOAD_DYLIB
          cmdsize 96
             name @rpath/libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.4.4.dylib.temp (offset 24)
       time stamp 2 Thu Jan  1 01:00:02 1970
          current version 0.0.0
    compatibility version 0.0.0
    ...
    Load command 17
              cmd LC_RPATH
          cmdsize 64
             path bazel-out/host/bin/tests/binary-with-lib-dynamic (offset 12)
    ```
    The corresponding rpath entry is a relative path. To my understanding this will cause the dynamic loader to look for the dylib relative to the execution path, not relative to the binary's path.

    Indeed, copying the `.dylib` to `.dylib.temp` in the same directory, by adding the following command to the `genrule`, does not change outcome.
    ```
    cp ./bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic.runfiles/io_tweag_rules_haskell/tests/binary-with-lib-dynamic/libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.4.4.dylib{,.temp}
    ```

    However, copying the dylib to the rpath relative to the execution directory, by adding the following command to the `genrule`, does prevent the dynamic linking error and the test-case passes.
    ```
    cp ./bazel-out/host/bin/tests/binary-with-lib-dynamic/binary-with-lib-dynamic.runfiles/io_tweag_rules_haskell/tests/binary-with-lib-dynamic/libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.4.4.dylib bazel-out/host/bin/tests/binary-with-lib-dynamic/libHStestsZSbinary-with-lib-dynamicZSlib-ghc8.4.4.dylib.temp
    ```

- `git bisect` shows that the issue was introduced by one of the following two commits: fdcf7ed1a2c4210b717bcf88e3905da84c4bd832 05fa76d33fe87ba34510912b19e1064d48ac3923

### PR DESCRIPTION
**Pull request by aherrmann**
- Fixes the dylib install name on MacOS. Previously, it included the `.temp` postfix, now it doesn't.
- Fixes the rpath on MacOS. Previously, it was relative to the execution directory. Now it is prefixed with `@loader_path` (analogous to `$ORIGIN` on Linux)
- Adds a regression test for #521.

Closes #521

### CODE REVIEW

File: haskell/private/actions/link.bzl [hunk 1]
```diff
@@ -43,6 +43,10 @@ def _fix_linker_paths(hs, inp, out, external_libraries):
             [
                 "cp {} {}".format(inp.path, out.path),
                 "chmod +w {}".format(out.path),
+                "/usr/bin/install_name_tool -id @rpath/{} {}".format(
```
Comment (Profpatsch) [inline haskell/private/actions/link.bzl]: According to the [manpage](https://www.manpagez.com/man/1/install_name_tool/) `-id` changes the “library identification name” of a dynamic shared library. I can’t find out what that is, Darwin’s [ld manpage](https://www.manpagez.com/man/1/ld/) doesn’t mention the word `identification` (or `id` for that matter).

Could you add an inline-comment what this line achieves? Would be awesome to have one for `-change` below as well.

File: haskell/private/actions/link.bzl [hunk 2]
```diff
@@ -227,11 +230,12 @@ def _add_external_libraries(args, libs):
                 "-L{0}".format(paths.dirname(lib.path)),
             ])
 
-def _infer_rpaths(target, solibs):
+def _infer_rpaths(hs, target, solibs):
     """Return set of RPATH values to be added to target so it can find all
     solibs.
 
     Args:
+      hs: Haskell context.
```
Comment (Profpatsch) [inline haskell/private/actions/link.bzl]: Adding `hs` is slightly overkill in this case, you could pass `is_darwin` directly.

File: haskell/private/actions/link.bzl (hunk 1 above)
Comment (aherrmann) [reply haskell/private/actions/link.bzl]: I've added some comments explaining this. The linked Oracle blog post also explains this to some extent.

File: haskell/private/actions/link.bzl (hunk 2 above)
Comment (aherrmann) [reply haskell/private/actions/link.bzl]: Done.

**Pull request merged**